server.py

```diff
@@ -23,53 +28,137 @@
+    if (os.path.exists(file_path)):
+        return send_from_directory(UPLOAD_FOLDER, secure_filename(filename))
+
+    redirect_address = metadata.lookup_file(filename, app.config['HOST'])
+    if (redirect_address is not None):
+        url = 'http://%s/read?%s' % (redirect_address, urllib.urlencode({ 'uuid': filename }))
+        return redirect(url, code=302)
+
+    other_servers = metadata.get_all_server(app.config['HOST'])
+    if (len(other_servers) > 0):
+        for server in other_servers:
+            url = 'http://%s/file_exists?%s' % (server, urllib.urlencode({ 'uuid': filename }))
+            lookup_request = requests.get(url)
+            if (lookup_request.status_code == 200):
+                redirection_url = 'http://%s/read?%s' % (server, urllib.urlencode({ 'uuid': filename }))
+                metadata.update_file_stored(filename, server)
+                metadata.close_connection()
+                return redirect(redirection_url, code=302)
+
+    return 'File Not Found', 404
+
+@app.route('/file_exists', methods=['GET'])
+def file_exists():
+    filename = request.args.get('uuid')
+    file_path = UPLOAD_FOLDER + secure_filename(filename)
+    print file_path
+    if (os.path.exists(file_path)):
+        return app.config['HOST'], 200
+    else:
+        return 'File not found', 404
 
+# Helper method for sending a file to another server
 def move_file(request):
     file_uuid = request.args.get('uuid')
     destination = request.args.get('destination')
     destination_with_endpoint = destination + '/write'
-    write_request = util.construct_put_request(destination_with_endpoint, file_uuid)
+    write_request = util.construct_post_request(destination_with_endpoint, file_uuid)
+    metadata = metadata_manager.MetadataManager()
+    metadata.update_file_stored(file_uuid, destination)
+    metadata.close_connection()
     return write_request
 
+# Transfers the file. This API call should not be open to all users.
 @app.route('/transfer', methods=['PUT'])
 def transfer():
     write_request = move_file(request)
     if write_request.status_code == 201:
         os.remove(os.path.join(app.config['UPLOAD_FOLDER'], file_uuid))
+        metadata.delete_file_stored(request.args.get('uuid'), request.args.get('destination'))
+        metadata.close_connection()
     return write_request
 
+# Replicate the file. This API call should not be open to all users.
 @app.route('/replicate', methods=['PUT'])
 def replicate():
     write_request = move_file(request)
     return write_request
 
+# Deletes the file. This API call should not be open to all users.
+@app.route('/delete', methods=['DELETE'])
+def delete():
+    file_uuid = request.args.get('uuid')
+    file_path = UPLOAD_FOLDER + '/' + file_uuid
+    if (os.path.exists(file_path)):
+        os.remove(file_path)
+        metadata = metadata_manager.MetadataManager()
+        metadata.delete_file_stored(file_uuid, app.config['HOST'])
```

Comment:
hm. maybe this should be outside the `os.path.exists` branch? reason is, if there is some bad record in the metadata, maybe you want to delete it even if the physical file doesn't exist.